File: receipt/score.py
"""Verifier v0.1.1 — verdict + trust score (FROZEN).

Implements docs/trust_score.md v0.1.1 EXACTLY. Server-side and client-side
(viewer/index.html) MUST produce identical numbers from the same input.

Public API:
    compute_score(events, report, window_start_ms=None, window_end_ms=None)
        -> dict matching docs/trust_score.md §3 output shape
"""
from __future__ import annotations

import time
from typing import Iterable

from receipt.chain import ChainStatus, VerifyReport, verify_chain

PERMANENT_ANCHORS = frozenset({"btc_op_return", "ethereum", "arweave"})
SEMI_PERMANENT_ANCHORS = frozenset({"ipfs"})
SOCIAL_ANCHORS = frozenset({"twitter", "moltbook", "github_commit"})
HEARTBEAT_BUCKET_MS = 5 * 60 * 1000  # 5 min
DEFAULT_WINDOW_MS = 7 * 24 * 3600 * 1000  # 7 days
SCHEMA_VERSION = "0.1.1"


def _anchor_tier(kind: str) -> str:
    if kind in PERMANENT_ANCHORS: return "permanent"
    if kind in SEMI_PERMANENT_ANCHORS: return "semi_permanent"
    if kind in SOCIAL_ANCHORS: return "social"
    return "unknown"


def _coverage(events: list[dict]) -> tuple[float, str, dict]:
    claims = {e["seq"] for e in events if e["kind"] == "claim"}
    if not claims:
        return 100.0, "no claims in window", {}
    terminated: set[int] = set()
    for e in events:
        cs = (e.get("data") or {}).get("claim_seq")
        if e["kind"] in ("verified", "error") and cs is not None:
            terminated.add(cs)
        elif e["kind"] == "signal_rejected":
            terminated.add(e["seq"])
    covered = len(claims & (terminated | {e["seq"] for e in events if e["kind"] == "signal_rejected"}))
    coverage = covered / len(claims)
    score = round(coverage * 100, 1)
    detail = f"{covered}/{len(claims)} claims have receipts"
    return score, detail, {"raw": coverage, "covered": covered, "total": len(claims)}


def _accuracy(events: list[dict]) -> tuple[float, str, dict]:
    verifs = [e for e in events if e["kind"] == "verified"]
    if not verifs:
        return 100.0, "no verified events to grade", {}
    receipts = len(verifs)
    verified_ok = sum(1 for e in verifs if (e["data"].get("match") or {}).get("id_match"))
    mismatches = receipts - verified_ok
    accuracy_raw = verified_ok / receipts
    mismatch_rate = mismatches / receipts
    score = max(0.0, (accuracy_raw - mismatch_rate * 2) * 100)  # 2× penalty per SPEC
    score = round(min(score, 100), 1)
    detail = f"{verified_ok}/{receipts} verified, {mismatches} mismatch{'es' if mismatches != 1 else ''}"
    return score, detail, {
        "verified_ok": verified_ok, "receipts": receipts,
        "mismatches": mismatches, "mismatch_rate": mismatch_rate,
    }


def _consistency(events: list[dict], window_start_ms: int, window_end_ms: int) -> tuple[float, str, dict]:
    """Consistency over an EXTERNAL window — caller-supplied or default
    last_7d_ending_at_chain_head. Never trust the chain's own [first, last]
    span: an attacker can omit early events to shrink the window and inflate
    density. SPEC §2.3."""
    span_ms = max(window_end_ms - window_start_ms, HEARTBEAT_BUCKET_MS)
    total_buckets = max(1, span_ms // HEARTBEAT_BUCKET_MS)
    covered: set[int] = set()
    for e in events:
        bucket = (e["ts"] - window_start_ms) // HEARTBEAT_BUCKET_MS
        if 0 <= bucket < total_buckets:
            covered.add(bucket)
    ratio = len(covered) / total_buckets
    score = round(min(ratio * 100, 100), 1)
    detail = f"covered {len(covered)}/{total_buckets} 5-min buckets in {(span_ms/3600000):.1f}h window"
    return score, detail, {"covered": len(covered), "total": total_buckets, "ratio": ratio,
                            "window_start_ms": window_start_ms, "window_end_ms": window_end_ms}


def _transparency(events: list[dict]) -> tuple[float, str, dict]:
    score = 100
    deductions: list[str] = []
    has_error = any(e["kind"] == "error" for e in events)
    has_recon = any(e["kind"] == "reconciliation" for e in events)
    has_tol = all(
        ((e["data"].get("match") or {}).get("tolerance_used")) is not None
        for e in events if e["kind"] == "verified"
    )
    if not has_error:
        score -= 20
        deductions.append("−20 no error events (real systems fail)")
    if not has_recon:
        score -= 30
        deductions.append("−30 no reconciliation event")
    if not has_tol:
        score -= 20
        deductions.append("−20 verified events missing match.tolerance_used")
    score = max(0, score)
    detail = "; ".join(deductions) if deductions else "all transparency criteria met"
    return float(score), detail, {"deductions": deductions}


def _integrity(events: list[dict], report: VerifyReport) -> tuple[float, str, dict]:
    """SPEC §2.5 — explicit anchor tiering: permanent / semi_permanent / social."""
    if report.status == ChainStatus.INVALID_CHAIN:
        return 0.0, "hash chain broken", {"chain_status": report.status.value}
    anchors = [e for e in events if e["kind"] == "anchor"]
    if not anchors:
        return 50.0, "no anchor event present", {"chain_status": report.status.value, "tier": None}
    last_anchor = anchors[-1]
    kind = last_anchor["data"].get("anchor_kind", "")
    tier = _anchor_tier(kind)
    last_event_ts = events[-1]["ts"]
    age_h = (last_event_ts - last_anchor["ts"]) / 3_600_000

    base = 100
    notes: list[str] = []
    if tier == "social":
        base = min(base, 90)
        notes.append(f"social tier ({kind}) — capped at 90")
    elif tier == "semi_permanent":
        base = min(base, 95)
        notes.append(f"semi-permanent tier ({kind}) — capped at 95")
    elif tier == "unknown":
        base = min(base, 80)
        notes.append(f"unknown anchor_kind ({kind}) — capped at 80")
    # else permanent — no cap

    if age_h > 24:
        base -= 30
        notes.append(f"anchor age {age_h:.1f}h (−30)")
    if age_h > 24 * 7:
        base = min(base, 30)
        notes.append("anchor > 7d (cap 30)")
    score = max(0, min(100, base))
    detail = f"anchor {age_h:.1f}h ago on {kind} [tier={tier}]" + (f"; {'; '.join(notes)}" if notes else "")
    return float(score), detail, {"age_hours": age_h, "tier": tier, "anchor_kind": kind}


def _flags(events: list[dict], dims: dict, accuracy_meta: dict) -> list[str]:
    flags: list[str] = []
    if accuracy_meta.get("mismatches", 0) > 0:
        flags.append(f"{accuracy_meta['mismatches']}_mismatches")
    error_count = sum(1 for e in events if e["kind"] == "error")
    if error_count:
        flags.append(f"{error_count}_error_events")
    if dims["Consistency"]["score"] < 60:
        flags.append("missing_heartbeat_periods")
    anchors = [e for e in events if e["kind"] == "anchor"]
    if not anchors:
        flags.append("no_anchor")
    elif anchors:
        age_h = (events[-1]["ts"] - anchors[-1]["ts"]) / 3_600_000
        if age_h > 24:
            flags.append("anchor_stale")
    return flags


def _suspicion_flags(events: list[dict], summary: dict, window_ms: int) -> list[str]:
    """SPEC §2.6 — additive flags, do NOT affect score. Surface human-intuition
    patterns the score can miss."""
    flags: list[str] = []

    # no_error_events — real systems fail; perfect logs are suspicious
    if summary.get("total_claims", 0) >= 20 and summary.get("errors", 0) == 0:
        flags.append("no_error_events")

    # sudden_activity_gap — any pair of consecutive events with >24h gap
    # (only relevant if window itself spans >24h)
    if window_ms > 24 * 3_600_000:
        max_gap_h = 0
        for a, b in zip(events, events[1:]):
            gap_h = (b["ts"] - a["ts"]) / 3_600_000
            max_gap_h = max(max_gap_h, gap_h)
        if max_gap_h > 24:
            flags.append("sudden_activity_gap")

    # high_mismatch_cluster — ≥3 mismatches in any rolling 10-verified window
    verifs = [e for e in events if e["kind"] == "verified"]
    if len(verifs) >= 10:
        for i in range(len(verifs) - 9):
            window = verifs[i:i + 10]
            mismatch = sum(1 for e in window if not (e["data"].get("match") or {}).get("id_match"))
            if mismatch >= 3:
                flags.append("mismatch_cluster")
                break

    # backfill_dominant_pretending_realtime — backfill > 5× realtime
    counts: dict[str, int] = {}
    for e in events:
        t = (e.get("data") or {}).get("trust_tier")
        if t:
            counts[t] = counts.get(t, 0) + 1
    backfill = counts.get("backfill_local", 0)
    realtime = counts.get("exchange_realtime", 0)
    if backfill > 0 and (realtime == 0 or backfill > realtime * 5):
        if realtime > 0:  # only flag if chain MIXES — pure-backfill is honest about itself
            flags.append("backfill_dominant_pretending_realtime")

    # anchor_only_social — every anchor in window is tier social
    anchors_in_window = [e for e in events if e["kind"] == "anchor"]
    if anchors_in_window and all(_anchor_tier(a["data"].get("anchor_kind", "")) == "social"
                                  for a in anchors_in_window):
        flags.append("only_social_anchor")

    # heartbeat_silence — window > 24h but no heartbeat
    if window_ms > 24 * 3_600_000 and not any(e["kind"] == "heartbeat" for e in events):
        flags.append("heartbeat_silence")

    return flags


def _examples(events: list[dict]) -> dict:
    """Concrete failure examples (SPEC §6 + Bonny's spec §六.examples)."""
    mismatches = []
    for e in events:
        if e["kind"] == "verified":
            m = e["data"].get("match") or {}
            if not m.get("id_match"):
                mismatches.append({"seq": e["seq"], "reason": "id_not_matched"})
            elif m.get("price_match") is False:
                mismatches.append({"seq": e["seq"], "reason": "price_out_of_tolerance"})
            elif m.get("size") is False:
                mismatches.append({"seq": e["seq"], "reason": "size_mismatch"})
    missing = []
    claim_seqs = {e["seq"] for e in events if e["kind"] == "claim"}
    terminated: set[int] = set()
    for e in events:
        cs = (e.get("data") or {}).get("claim_seq")
        if e["kind"] in ("verified", "error") and cs is not None:
            terminated.add(cs)
        elif e["kind"] == "signal_rejected":
            terminated.add(e["seq"])
    for s in sorted(claim_seqs - terminated)[:5]:
        missing.append({"seq": s})
    return {"mismatch": mismatches[:5], "missing_receipts": missing}


def _trust_tier_majority(events: list[dict]) -> str:
    """Return the predominant trust_tier on verified+reconciliation events."""
    counts: dict[str, int] = {}
    for e in events:
        if e["kind"] in ("verified", "reconciliation"):
            t = (e.get("data") or {}).get("trust_tier")
            if t:
                counts[t] = counts.get(t, 0) + 1
    if not counts:
        return "manual"
    return max(counts, key=counts.get)


def _verdict(coverage: dict, accuracy: dict, has_anchor: bool, has_recon: bool,
             anchor_recent: bool, consistency_score: float, tier_majority: str) -> tuple[str, str]:
    """Rule-based per docs/trust_score.md §1.

    Verdict cap by trust_tier majority — a chain made of backfill_local can
    at best be Mixed; a chain made of manual at best Unverified. This
    prevents 'I made up the receipts but the math passes' from going green.
    """
    cov = coverage.get("raw", 0)
    mismatch_rate = accuracy.get("mismatch_rate", 0)
    accuracy_raw = (accuracy.get("verified_ok", 0) / accuracy["receipts"]) if accuracy.get("receipts") else 1.0

    realtime_tiers = {"exchange_realtime", "exchange_delayed", "api_verified"}
    qualifies_for_verified = (
        cov >= 0.9 and accuracy_raw >= 0.8 and mismatch_rate <= 0.05
        and has_anchor and has_recon and anchor_recent
        and tier_majority in realtime_tiers
    )
    fails_hard = (
        cov < 0.5 or mismatch_rate > 0.20
        or not has_anchor or consistency_score < 30
        or tier_majority == "manual"
    )

    if qualifies_for_verified:
        return "Verified", "green"
    if fails_hard:
        return "Unverified", "red"
    return "Mixed", "yellow"


def _enforce_verdict_score_consistency(verdict: str, score: int) -> int:
    """SPEC §4 — verdict and score must not contradict."""
    if verdict == "Verified" and score < 75:
        return 75
    if verdict == "Unverified" and score >= 40:
        return 39
    if verdict == "Mixed":
        return max(40, min(74, score))
    return score


def compute_score(events: Iterable[dict], report: VerifyReport | None = None,
                  window_start_ms: int | None = None,
                  window_end_ms: int | None = None) -> dict:
    """Return verdict + trust score per docs/trust_score.md §3 output shape.

    Window semantics (SPEC §2.3):
        - If both window_start_ms and window_end_ms supplied → use them
        - Else default to last 7 days ending at chain head (or chain span if
          chain is shorter than 7 days)
        - Window is recorded in output for transparency
    """
    events = list(events)
    if report is None:
        report = verify_chain(events)
    if not events:
        return {
            "agent": None, "verdict": "Unverified", "verdict_color": "red",
            "score": 0, "dimensions": {}, "summary": {},
            "rule_breaks": ["empty"], "suspicion_flags": [],
            "examples": {"mismatch": [], "missing_receipts": []},
            "schema_version": SCHEMA_VERSION,
            "chain_status": report.status.value,
            "computed_at_ms": int(time.time() * 1000),
        }
    # SPEC §1: INVALID_CHAIN is fatal — score=0, verdict=Unverified, no
    # further dimensional computation. Mirrors viewer/score.js short-circuit.
    if report.status == ChainStatus.INVALID_CHAIN:
        # Match viewer/score.js INVALID_CHAIN shape exactly: no raw_score,
        # no trust_tier_majority, no window — INVALID is terminal.
        return {
            "agent": events[0].get("agent"),
            "schema_version": SCHEMA_VERSION,
            "verdict": "Unverified", "verdict_color": "red",
            "score": 0,
            "dimensions": {"Integrity": {"score": 0, "detail": "hash chain broken"}},
            "summary": {}, "rule_breaks": ["invalid_chain"],
            "suspicion_flags": [],
            "examples": {"mismatch": [], "missing_receipts": []},
            "chain_status": report.status.value,
            "computed_at_ms": int(time.time() * 1000),
        }

    # Resolve window — caller-supplied OR last 7d from chain head.
    # Critical: do NOT clip to chain_first_ts. Per SPEC §2.3, the window must
    # be externally-anchored, otherwise an attacker omits early events to
    # shrink the window and inflate density. New chains < 7d old correctly
    # show lower Consistency until they accumulate real history.
    chain_head_ts = events[-1]["ts"]
    if window_end_ms is None:
        window_end_ms = chain_head_ts
    if window_start_ms is None:
        window_start_ms = window_end_ms - DEFAULT_WINDOW_MS
    window_ms = window_end_ms - window_start_ms

    cov_score, cov_detail, cov_meta = _coverage(events)
    acc_score, acc_detail, acc_meta = _accuracy(events)
    cons_score, cons_detail, cons_meta = _consistency(events, window_start_ms, window_end_ms)
    trans_score, trans_detail, trans_meta = _transparency(events)
    int_score, int_detail, int_meta = _integrity(events, report)

    raw_score = round((cov_score + acc_score + cons_score + trans_score + int_score) / 5)

    has_anchor = any(e["kind"] == "anchor" for e in events)
    has_recon = any(e["kind"] == "reconciliation" for e in events)
    anchor_recent = False
    if has_anchor:
        anchors = [e for e in events if e["kind"] == "anchor"]
        anchor_recent = (events[-1]["ts"] - anchors[-1]["ts"]) <= 24 * 3_600_000

    tier_majority = _trust_tier_majority(events)
    verdict, color = _verdict(cov_meta, acc_meta, has_anchor, has_recon, anchor_recent,
                              cons_score, tier_majority)
    score = _enforce_verdict_score_consistency(verdict, raw_score)

    dims = {
        "Coverage":     {"score": cov_score,   "detail": cov_detail},
        "Accuracy":     {"score": acc_score,   "detail": acc_detail},
        "Consistency":  {"score": cons_score,  "detail": cons_detail},
        "Transparency": {"score": trans_score, "detail": trans_detail},
        "Integrity":    {"score": int_score,   "detail": int_detail},
    }

    summary = {
        "total_claims": cov_meta.get("total", 0),
        "receipts": acc_meta.get("receipts", 0),
        "verified": acc_meta.get("verified_ok", 0),
        "mismatch": acc_meta.get("mismatches", 0),
        "errors": sum(1 for e in events if e["kind"] == "error"),
    }
    return {
        "agent": events[0].get("agent"),
        "schema_version": SCHEMA_VERSION,
        "window": {"from_ms": window_start_ms, "to_ms": window_end_ms,
                   "source": "caller" if (window_end_ms != chain_head_ts) else "default_last_7d"},
        "verdict": verdict,
        "verdict_color": color,
        "score": score,
        "raw_score_pre_clamp": raw_score,
        "trust_tier_majority": tier_majority,
        "dimensions": dims,
        "summary": summary,
        "rule_breaks": _flags(events, dims, acc_meta),
        "suspicion_flags": _suspicion_flags(events, summary, window_ms),
        "examples": _examples(events),
        "chain_status": report.status.value,
        "computed_at_ms": int(time.time() * 1000),
    }

File: viewer/score.js
// Receipt v0.1.1 — verifier algorithm (browser + Node compatible).
//
// Single source of truth for the JS implementation. Used by:
//   - viewer/index.html (browser via <script src>)
//   - tests/cross_impl_test.py (Node, via subprocess)
//
// Cross-impl invariant: this file + receipt/score.py + receipt/chain.py
// MUST produce byte-identical outputs for any input chain. Tests/cross_impl
// enforces this on fixtures.
//
// Hashing: detects environment.
//   - Browser: globalThis.crypto.subtle.digest (async)
//   - Node 16+: node:crypto.createHash (sync)

(function (root, factory) {
  if (typeof module === "object" && module.exports) {
    module.exports = factory(require("crypto"));  // Node
  } else {
    root.ReceiptScore = factory(null);            // Browser
  }
}(typeof self !== "undefined" ? self : this, function (nodeCrypto) {

  const SCORE_SCHEMA_VERSION = "0.1.1";
  const HASH_PREFIX = "sha256:";
  const GENESIS_PREV = HASH_PREFIX + "0".repeat(64);
  const CANONICAL_KEYS = ["v","schema_version","ts","seq","agent","kind","data","prev_hash"];
  const PERMANENT_ANCHORS_TIER = new Set(["btc_op_return","ethereum","arweave"]);
  const SEMI_PERM_ANCHORS_TIER = new Set(["ipfs"]);
  const SOCIAL_ANCHORS_TIER = new Set(["twitter","moltbook","github_commit"]);
  const REALTIME_TIERS = new Set(["exchange_realtime","exchange_delayed","api_verified"]);
  const HEARTBEAT_BUCKET_MS = 5*60*1000;
  const DEFAULT_WINDOW_MS = 7*24*3600*1000;

  function anchorTier(kind) {
    if (PERMANENT_ANCHORS_TIER.has(kind)) return "permanent";
    if (SEMI_PERM_ANCHORS_TIER.has(kind)) return "semi_permanent";
    if (SOCIAL_ANCHORS_TIER.has(kind)) return "social";
    return "unknown";
  }

  // ── Hashing (env-detected) ───────────────────────────────────────────
  async function sha256hex(text) {
    if (nodeCrypto) {
      return nodeCrypto.createHash("sha256").update(text, "utf8").digest("hex");
    }
    const buf = new TextEncoder().encode(text);
    const hash = await globalThis.crypto.subtle.digest("SHA-256", buf);
    return [...new Uint8Array(hash)].map(b => b.toString(16).padStart(2,"0")).join("");
  }

  function jsonCanonical(v) {
    if (v === null) return "null";
    if (typeof v === "boolean") return v ? "true" : "false";
    if (typeof v === "number") return JSON.stringify(v);
    if (typeof v === "string") return JSON.stringify(v);
    if (Array.isArray(v)) return "[" + v.map(jsonCanonical).join(",") + "]";
    return "{" + Object.keys(v).sort().map(k => JSON.stringify(k)+":"+jsonCanonical(v[k])).join(",") + "}";
  }

  function canonical(ev) {
    const body = {};
    for (const k of CANONICAL_KEYS) if (k in ev) body[k] = ev[k];
    return jsonCanonical(body);
  }

  // ── verifyChain ───────────────────────────────────────────────────────
  async function verifyChain(events) {
    if (events.length === 0) return {status:"EMPTY", n_events:0, n_claims:0, n_verified:0, n_errors:0, n_rejected:0, issues:[]};
    let prev = GENESIS_PREV, expectedSeq = 0;
    const claims=new Set(), verified=new Set(), errors=new Set(), rejected=new Set(), suspect=new Set();
    let lastReconTs=null, lastEventTs=null;
    const issues = [];
    for (const ev of events) {
      if (ev.seq !== expectedSeq) { issues.push("seq gap at " + expectedSeq); return fail(); }
      if (ev.prev_hash !== prev) { issues.push("prev_hash mismatch at seq " + expectedSeq); return fail(); }
      const recomputed = HASH_PREFIX + (await sha256hex(canonical(ev)));
      if (ev.hash !== recomputed) { issues.push("hash mismatch at seq " + expectedSeq); return fail(); }
      const d = ev.data || {};
      if (ev.kind === "claim") claims.add(ev.seq);
      else if (ev.kind === "verified") {
        if (d.claim_seq != null) verified.add(d.claim_seq);
        if (!(d.match || {}).id_match) suspect.add(ev.seq);
      } else if (ev.kind === "error") { if (d.claim_seq != null) errors.add(d.claim_seq); }
      else if (ev.kind === "signal_rejected") rejected.add(ev.seq);
      else if (ev.kind === "reconciliation") lastReconTs = ev.ts;
      lastEventTs = ev.ts; prev = ev.hash; expectedSeq++;
    }
    function fail() { return {status:"INVALID_CHAIN", n_events:events.length, n_claims:0, n_verified:0, n_errors:0, n_rejected:0, issues}; }
    return {
      status: "OK",
      n_events: events.length, n_claims: claims.size,
      n_verified: [...claims].filter(c => verified.has(c)).length,
      n_errors: errors.size, n_rejected: rejected.size, issues,
    };
  }

  function trustTierMajority(events) {
    const counts = {};
    for (const e of events) {
      if (e.kind === "verified" || e.kind === "reconciliation") {
        const t = (e.data||{}).trust_tier;
        if (t) counts[t] = (counts[t]||0)+1;
      }
    }
    if (Object.keys(counts).length === 0) return "manual";
    return Object.entries(counts).sort((a,b)=>b[1]-a[1])[0][0];
  }

  // ── computeScore ──────────────────────────────────────────────────────
  function computeScore(events, report) {
    if (events.length === 0) return {schema_version: SCORE_SCHEMA_VERSION, verdict:"Unverified", verdict_color:"red", score:0, dimensions:{}, summary:{}, rule_breaks:["empty"], suspicion_flags:[], examples:{mismatch:[],missing_receipts:[]}};
    if (report.status === "INVALID_CHAIN") return {schema_version: SCORE_SCHEMA_VERSION, verdict:"Unverified", verdict_color:"red", score:0, dimensions:{Integrity:{score:0,detail:"hash chain broken"}}, summary:{}, rule_breaks:["invalid_chain"], suspicion_flags:[], examples:{mismatch:[],missing_receipts:[]}};

    const claims = new Set(events.filter(e=>e.kind==="claim").map(e=>e.seq));
    const terminated = new Set();
    for (const e of events) {
      const cs = (e.data||{}).claim_seq;
      if ((e.kind==="verified"||e.kind==="error") && cs!=null) terminated.add(cs);
      else if (e.kind==="signal_rejected") terminated.add(e.seq);
    }
    const covered = [...claims].filter(c=>terminated.has(c)).length;
    const coverage = claims.size === 0 ? 1 : covered / claims.size;
    const cov_score = +(coverage * 100).toFixed(1);
    const cov_detail = `${covered}/${claims.size||0} claims have receipts`;

    const verifs = events.filter(e=>e.kind==="verified");
    const verified_ok = verifs.filter(e => (e.data.match||{}).id_match).length;
    const receipts = verifs.length;
    const mismatches = receipts - verified_ok;
    const mismatch_rate = receipts ? mismatches / receipts : 0;
    const accuracy_raw = receipts ? verified_ok / receipts : 1;
    const acc_score = +Math.max(0, Math.min(100, (accuracy_raw - mismatch_rate*2)*100)).toFixed(1);
    const acc_detail = receipts ? `${verified_ok}/${receipts} verified, ${mismatches} mismatch${mismatches!==1?'es':''}` : "no verified events";

    const last = events[events.length-1].ts;
    const windowEnd = last;
    const windowStart = windowEnd - DEFAULT_WINDOW_MS;
    const span = Math.max(windowEnd - windowStart, HEARTBEAT_BUCKET_MS);
    const totalBuckets = Math.max(1, Math.floor(span/HEARTBEAT_BUCKET_MS));
    const buckets = new Set();
    for (const e of events) {
      const b = Math.floor((e.ts - windowStart) / HEARTBEAT_BUCKET_MS);
      if (b >= 0 && b < totalBuckets) buckets.add(b);
    }
    const cons_ratio = buckets.size / totalBuckets;
    const cons_score = +Math.min(cons_ratio*100, 100).toFixed(1);
    const cons_detail = `covered ${buckets.size}/${totalBuckets} 5-min buckets in ${(span/3600000).toFixed(0)}h window`;

    let trans_score = 100;
    const trans_notes = [];
    if (!events.some(e=>e.kind==="error")) { trans_score -= 20; trans_notes.push("−20 no error events (real systems fail)"); }
    if (!events.some(e=>e.kind==="reconciliation")) { trans_score -= 30; trans_notes.push("−30 no reconciliation event"); }
    const allHaveTol = verifs.every(e => ((e.data.match||{}).tolerance_used) != null);
    if (!allHaveTol) { trans_score -= 20; trans_notes.push("−20 missing match.tolerance_used"); }
    trans_score = Math.max(0, trans_score);
    const trans_detail = trans_notes.length ? trans_notes.join("; ") : "all transparency criteria met";

    const anchors = events.filter(e=>e.kind==="anchor");
    let int_score = 100, int_detail = "chain valid";
    if (anchors.length === 0) { int_score = 50; int_detail = "no anchor event"; }
    else {
      const a = anchors[anchors.length-1];
      const kind = a.data.anchor_kind || "";
      const tier = anchorTier(kind);
      const ageH = (last - a.ts)/3600000;
      const notes = [];
      if (tier === "social") { int_score = Math.min(int_score, 90); notes.push(`social tier (${kind}) capped 90`); }
      else if (tier === "semi_permanent") { int_score = Math.min(int_score, 95); notes.push(`semi-permanent (${kind}) capped 95`); }
      else if (tier === "unknown") { int_score = Math.min(int_score, 80); notes.push(`unknown anchor (${kind}) capped 80`); }
      if (ageH > 24) { int_score -= 30; notes.push(`anchor ${ageH.toFixed(1)}h old (−30)`); }
      if (ageH > 168) { int_score = Math.min(int_score, 30); notes.push("anchor > 7d cap 30"); }
      int_score = Math.max(0, Math.min(100, int_score));
      int_detail = `anchor ${ageH.toFixed(1)}h ago on ${kind} [tier=${tier}]` + (notes.length ? "; " + notes.join("; ") : "");
    }

    const dims = {
      Coverage:     {score: cov_score,   detail: cov_detail},
      Accuracy:     {score: acc_score,   detail: acc_detail},
      Consistency:  {score: cons_score,  detail: cons_detail},
      Transparency: {score: trans_score, detail: trans_detail},
      Integrity:    {score: int_score,   detail: int_detail},
    };
    const raw = Math.round((cov_score+acc_score+cons_score+trans_score+int_score)/5);

    const tierMaj = trustTierMajority(events);
    const hasAnchor = anchors.length > 0;
    const hasRecon = events.some(e=>e.kind==="reconciliation");
    const anchorRecent = hasAnchor && (last - anchors[anchors.length-1].ts) <= 24*3600000;
    let verdict, color;
    if (coverage >= 0.9 && accuracy_raw >= 0.8 && mismatch_rate <= 0.05
        && hasAnchor && hasRecon && anchorRecent && REALTIME_TIERS.has(tierMaj)) {
      verdict = "Verified"; color = "green";
    } else if (coverage < 0.5 || mismatch_rate > 0.20 || !hasAnchor || cons_score < 30 || tierMaj === "manual") {
      verdict = "Unverified"; color = "red";
    } else {
      verdict = "Mixed"; color = "yellow";
    }

    let score = raw;
    if (verdict === "Verified" && score < 75) score = 75;
    if (verdict === "Unverified" && score >= 40) score = 39;
    if (verdict === "Mixed") score = Math.max(40, Math.min(74, score));

    const errCount = events.filter(e=>e.kind==="error").length;
    const flags = [];
    if (mismatches > 0) flags.push(`${mismatches}_mismatches`);
    if (errCount) flags.push(`${errCount}_error_events`);
    if (cons_score < 60) flags.push("missing_heartbeat_periods");
    if (anchors.length === 0) flags.push("no_anchor");
    else if ((last - anchors[anchors.length-1].ts)/3600000 > 24) flags.push("anchor_stale");
    if (tierMaj === "backfill_local") flags.push("backfill_only_evidence");
    if (tierMaj === "manual") flags.push("manual_attested_only");

    const exMismatch = [];
    for (const e of verifs) {
      const m = e.data.match || {};
      if (!m.id_match) exMismatch.push({seq:e.seq, reason:"id_not_matched"});
      else if (m.price_match === false) exMismatch.push({seq:e.seq, reason:"price_out_of_tolerance"});
      else if (m.size === false) exMismatch.push({seq:e.seq, reason:"size_mismatch"});
    }
    const exMissing = [];
    for (const c of [...claims].filter(c=>!terminated.has(c)).slice(0,5)) exMissing.push({seq:c});

    const suspicion = [];
    if (claims.size >= 20 && errCount === 0) suspicion.push("no_error_events");
    if (DEFAULT_WINDOW_MS > 24*3600000) {
      let maxGap = 0;
      for (let i=1; i<events.length; i++) maxGap = Math.max(maxGap, (events[i].ts - events[i-1].ts)/3600000);
      if (maxGap > 24) suspicion.push("sudden_activity_gap");
    }
    if (verifs.length >= 10) {
      for (let i=0; i<=verifs.length-10; i++) {
        const win = verifs.slice(i, i+10);
        const mm = win.filter(e => !((e.data.match||{}).id_match)).length;
        if (mm >= 3) { suspicion.push("mismatch_cluster"); break; }
      }
    }
    const tierCounts = {};
    for (const e of events) {
      const t = (e.data||{}).trust_tier;
      if (t) tierCounts[t] = (tierCounts[t]||0)+1;
    }
    const bf = tierCounts.backfill_local||0, rt = tierCounts.exchange_realtime||0;
    if (bf > 0 && rt > 0 && bf > rt*5) suspicion.push("backfill_dominant_pretending_realtime");
    if (anchors.length > 0 && anchors.every(a => anchorTier(a.data.anchor_kind||"") === "social")) suspicion.push("only_social_anchor");
    if (DEFAULT_WINDOW_MS > 24*3600000 && !events.some(e=>e.kind==="heartbeat")) suspicion.push("heartbeat_silence");

    return {
      schema_version: SCORE_SCHEMA_VERSION,
      window: {from_ms: windowStart, to_ms: windowEnd, source: "default_last_7d"},
      verdict, verdict_color: color, score, raw_score_pre_clamp: raw,
      trust_tier_majority: tierMaj,
      dimensions: dims,
      summary: {total_claims: claims.size, receipts, verified: verified_ok, mismatch: mismatches, errors: errCount},
      rule_breaks: flags,
      suspicion_flags: suspicion,
      examples: {mismatch: exMismatch.slice(0,5), missing_receipts: exMissing},
    };
  }

  return { verifyChain, computeScore, anchorTier, SCORE_SCHEMA_VERSION };
}));
